t/ingredients_percent.t

```diff
@@ -0,0 +1,349 @@
+          }
+        ],
+        'percent_max' => 100,
+        'percent_min' => 70,
+        'text' => 'cocoa'
+      },
+      {
+        'id' => 'en:sugar',
+        'percent_max' => 30,
+        'percent_min' => 0,
+        'text' => 'sugar'
+      }
+    ],
+    'percent_max' => 100,
+    'percent_min' => 100,
+    'text' => 'chocolate'
+  }
+
+]
+	],
+
+# Make sure we can handle impossible values gracefully
+
+# This ingredient string caused an infinite loop:
+#  "farine (12%), chocolat (beurre de cacao (15%), sucre [10%], protéines de lait, oeuf 1%) - émulsifiants : E463, E432 et E472 - correcteurs d'acidité : E322/E333 E474-E475, acidifiant (acide citrique, acide phosphorique) - sel"
+
+        [ { lc => "fr", ingredients_text => "beurre de cacao (15%), sucre [10%], protéines de lait, oeuf 1%"},
+[
+  {
+    'id' => 'en:cocoa-butter',
+    'percent' => '15',
+    'percent_max' => 15,
+    'percent_min' => 15,
+    'text' => 'beurre de cacao'
+  },
+  {
+    'id' => 'en:sugar',
+    'percent' => '10',
+    'percent_max' => '10',
+    'percent_min' => '10',
+    'text' => 'sucre'
+  },
+  {
+    'id' => 'en:milk-proteins',
+    'percent_max' => 100,
+    'percent_min' => 0,
+    'text' => "prot\x{e9}ines de lait"
+  },
+  {
+    'id' => 'en:egg',
+    'percent' => '1',
+    'percent_max' => '1',
+    'percent_min' => '1',
+    'text' => 'oeuf'
+  }
+
+]
+	],
+
+        [ { lc => "fr", ingredients_text => "farine (12%), chocolat (beurre de cacao (15%), sucre [10%], protéines de lait, oeuf 1%)"},
+[
+  {
+    'id' => 'en:flour',
+    'percent' => '12',
+    'percent_max' => 12,
+    'percent_min' => 12,
+    'text' => 'farine'
+  },
+  {
+    'id' => 'en:chocolate',
+    'ingredients' => [
+      {
+        'id' => 'en:cocoa-butter',
+        'percent' => '15',
+        'text' => 'beurre de cacao'
+      },
+      {
+        'id' => 'en:sugar',
+        'percent' => '10',
+        'text' => 'sucre'
+      },
+      {
+        'id' => 'en:milk-proteins',
+        'text' => "prot\x{e9}ines de lait"
+      },
+      {
+        'id' => 'en:egg',
+        'percent' => '1',
+        'text' => 'oeuf'
+      }
+    ],
+    'percent_max' => 100,
+    'percent_min' => 0,
+    'text' => 'chocolat'
+  }
+
+]
+        ],
+
+);
+
+foreach my $test_ref (@tests) {
+
+	my $product_ref = $test_ref->[0];
+	my $expected_ingredients_ref = $test_ref->[1];
+
+	print STDERR "ingredients_text: " . $product_ref->{ingredients_text} . "\n";
+
+	parse_ingredients_text($product_ref);
+	compute_ingredients_percent_values(
+		100, 100, $product_ref->{ingredients});
+
+	is_deeply ($product_ref->{ingredients}, $expected_ingredients_ref)
+		# using print + join instead of diag so that we don't have
+		# hashtags. It makes copy/pasting the resulting structure
+		# inside the test file much easier when tests results need
+		# to be updated. Caveat is that it might interfere with
+		# test output.
+		or print STDERR join("\n", explain $product_ref->{ingredients});
```

Comment:
`print STDERR` isn't recommended in unit tests: https://stackoverflow.com/a/6162805/11963